### PR TITLE
add ApplyPlanningSceneService capability

### PR DESCRIPTION
**Pull request by v4hn**
See https://groups.google.com/forum/#!topic/moveit-users/D7MWZEUSKLc
for a longer discussion on why this makes sense to have around.

The only thing I'm unsure about is whether the service should be
`~`-namespaced or not. There might be multiple monitored scenes
around, so I moved it to the namespace. On the other hand,
the ClearOctomap service is _not_ namespaced...

### CODE REVIEW

**Comment by v4hn** [comment]
This obviously depends on https://github.com/ros-planning/moveit_msgs/pull/21

File: move_group/src/default_capabilities/apply_planning_scene_service_capability.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Michael 'v4hn' Goerner */
```
Comment (davetcoleman) [inline move_group/src/default_capabilities/apply_planning_scene_service_capability.h:35]: im a big fan of descriptions in header files for what the class does....

File: planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h [hunk 2]
```diff
@@ -478,9 +481,6 @@ class PlanningSceneMonitor : private boost::noncopyable
   // Callback for a new planning scene msg
   void newPlanningSceneCallback(const moveit_msgs::PlanningSceneConstPtr &scene);
 
-  // Called to update the planning scene with a new message.
-  void newPlanningSceneMessage(const moveit_msgs::PlanningScene& scene);
```
Comment (davetcoleman) [inline planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h:482]: just a guess, but wouldn't this break ABI too? I don't care about ABI, actually, but maybe we should go ahead and make it return bool while we're at it (doesn't break API).

File: move_group/src/default_capabilities/apply_planning_scene_service_capability.h (hunk 1 above)
Comment (v4hn) [inline move_group/src/default_capabilities/apply_planning_scene_service_capability.h:35]: On Mon, Jun 13, 2016 at 03:41:28PM -0700, Dave Coleman wrote:

> im a big fan of descriptions in header files for what the class does....

Now you're being unfair. :)
I totally agree, but none of the other capabilities include a description
and I wouldn't even know what to describe here, as it's pretty self-documenting.
"ApplyPlanningSceneService : MoveGroupCapability is a MoveGroupCapability
that provides a service to apply a given PlanningScene" is just overly redundant.

File: planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h (hunk 2 above)
Comment (v4hn) [inline planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h:482]: On Mon, Jun 13, 2016 at 03:42:43PM -0700, Dave Coleman wrote:

> just a guess, but wouldn't this break ABI too?

No, changing visibility private -> public does not break ABI.
As it turns out, in this case changing the `void` return
to `bool` also is ABI compliant as the function was private before.
I'll prepare the changes.

**Comment by v4hn** [comment]
@davetcoleman: now this request also spans 4 repos :-)
Full +1 to merge most moveit repositories...

I do care quite a lot about a compatible ABI within ros distros until we add sonames to our libraries to explicitly mark updates as ABI-breaking.
As @mikeferguson pointed out last week it's pretty horrible to debug random segfaults due to ABI changes without sonames. Last year I spent more than 12 hours debugging one of these issues in the PCL, although it was a particularly horrible one.

Any comments on the question of namespacing?

File: move_group/src/default_capabilities/apply_planning_scene_service_capability.h (hunk 1 above)
Comment (davetcoleman) [inline move_group/src/default_capabilities/apply_planning_scene_service_capability.h:35]: You are exactly right about not having enough documentation throughout the code base and I know the others lack it. How about:
"Provides ability to update the shared planning scene with a remote blocking call"

File: planning/planning_scene_monitor/src/planning_scene_monitor.cpp [hunk 3]
```diff
@@ -476,73 +476,77 @@ void planning_scene_monitor::PlanningSceneMonitor::clearOctomap()
...
 }
 
-void planning_scene_monitor::PlanningSceneMonitor::newPlanningSceneMessage(const moveit_msgs::PlanningScene& scene)
+bool planning_scene_monitor::PlanningSceneMonitor::newPlanningSceneMessage(const moveit_msgs::PlanningScene& scene)
 {
-  if (scene_)
+  if (!scene_)
+    return false;
+
+  bool ret;
```
Comment (davetcoleman) [inline planning/planning_scene_monitor/src/planning_scene_monitor.cpp]: result

File: planning/planning_scene_monitor/src/planning_scene_monitor.cpp [hunk 4]
```diff
@@ -476,73 +476,77 @@ void planning_scene_monitor::PlanningSceneMonitor::clearOctomap()
...
+
+  SceneUpdateType upd = UPDATE_SCENE;
```
Comment (davetcoleman) [inline planning/planning_scene_monitor/src/planning_scene_monitor.cpp:486]: again, full var names...

**Comment by davetcoleman** [comment]
i don't have an opinion on the namespacing, other than we should do it as the other services have been done. i don't know of any current uses cases needing more than one planning scene

**Comment by v4hn** [comment]
Ok, I renamed all the `result` variables,
added the comment you proposed, and decided to move
the service away from `~`. Most of the other services
don't use it so no reason to start with it for now.

**Comment by davetcoleman** [comment]
I'll merge these as soon as @mikeferguson gives me a +1 on the moveit_msgs change. Until then, could you add a section to the documentation [here](http://docs.ros.org/indigo/api/pr2_moveit_tutorials/html/planning/src/doc/planning_scene_ros_api_tutorial.html) mentioning this new way to update the PS? It doesn't need to be much.

**Comment by mikeferguson** [comment]
@davetcoleman -- we need to merge the moveit_msgs and then release it as Debs so that this can properly build, else the CI will be broken

**Comment by davetcoleman** [comment]
just restarted build after merging msgs and core

**Pull request merged**